File: 23-merge-k-sorted-lists/merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def mergeTwoLists(self, list1, list2):

        head = ListNode(None)
        p = head

        while list1 and list2:
            if list1.val <= list2.val:
                p.next = list1
                list1 = list1.next
            else:
                p.next = list2
                list2 = list2.next
            p = p.next

        p.next = list1 if list1 else list2

        return head.next

    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        if not lists:
            return None
            
        if len(lists) == 1:
            return lists[0]

        lists_len = len(lists)
        res1 = self.mergeKLists(lists[0:lists_len / 2])
        res2 = self.mergeKLists(lists[lists_len / 2:])
        return self.mergeTwoLists(res1, res2)
        

File: 23-merge-k-sorted-lists/merge-k-sorted-lists.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func mergeTwoLists(l1, l2 *ListNode) *ListNode {
	if l1 == nil {
		return l2
	}
	if l2 == nil {
		return l1
	}

	head := &ListNode{}
	p := head
	for l1 != nil && l2 != nil {
		if l1.Val > l2.Val {
			p.Next = l2
			l2 = l2.Next
		} else {
			p.Next = l1
			l1 = l1.Next
		}
		p = p.Next
	}

	if l1 == nil {
		p.Next = l2
	} else {
		p.Next = l1
	}

	return head.Next
}

func mergeKLists(lists []*ListNode) *ListNode {
    if lists == nil || len(lists) == 0 {
        return nil
    }
    if len(lists) == 1 {
        return lists[0]
    }

    l1 := mergeKLists(lists[0:len(lists)/2])
    l2 := mergeKLists(lists[len(lists)/2:])

    return mergeTwoLists(l1, l2)
    
}